### ISSUE TITLE
Lz4FrameEncoder (and other compressors) Acknowledge Writes Too Early

### ISSUE THREAD

**Issue by belliottsmith** [netty/netty#9281]
Expected behaviour: The promise for a write should be completed only once the write makes it to the network

Actual behaviour: Buffering encoders that extend `MessageToByteEncoder`, most notably Lz4FrameEncoder and other compressors, will notify the promise before the frame containing the message reaches the network.

This behaviour is triggered by `MessageToByteEncoder.write` which is designed for 1:1 conversions from message to `ByteBuf`.  For the affected compressors, the `encode` method returns a `ByteBuf` that does not fully contain the message, unless it happens to precisely fill the buffer.  

Either an empty `ByteBuf` is returned, or one containing only part of the message, with the remainder in both cases left in the buffer.  In either case, the supplied promise will be attached to the write of this `ByteBuf` (empty or otherwise), and not to the `ByteBuf` that contains the end of the message.

Netty version: 4.0+ (I think, certainly confirmed on 4.1)

**Comment by belliottsmith** [comment]
Unfortunately resolving this issue is less trivial than we might hope, as each buffering `MessageToByteEncoder` needs to be updated.  We probably want to introduce a `MessageToBufferedByteEncoder` or some such, that also buffers promises for all messages in a given buffer.  This will mean updating a lot of classes that extend `MessageToByteEncoder`.

Note that the `SslHandler` effectively already handles this by buffering the unencrypted messages alongside their promise.  Whether or not we want to port the `SslHandler` to any new parent class is also a consideration.

**Comment by njhill** [comment]
@belliottsmith interesting finding :) It does look like a mis-use of `MessageToByteEncoder`, which doesn't contain much logic to begin with. I like your idea of introducing a new `MessageToBufferedByteEncoder` (`BufferingMessageToByteEncoder`?) - hopefully it could be a non-breaking addition if it extended `MessageToByteEncoder`.

It could have an alternative `encode` method (`bufferedEncode`?) which returns a `boolean` indicating whether the input message was fully included in `out`, and override `encode` just to call this. Then existing impls would hopefully need minimal change, i.e. just rename their `encode` and return the appropriate val instead of `void`. WDYT?

I'm tempted to label this issue `bug` but will leave that to @normanmaurer in case there is something we're missing...

**Comment by normanmaurer** [comment]
@njhill yes it is a bug and yes I think we should have a new base-class that extends `MessageToByteEncoder` for this. This way we can hopefully implement it without breaking the API (which we can't for 4.1).

**Comment by normanmaurer** [comment]
@belliottsmith thanks for bringing this up.

**Comment by belliottsmith** [comment]
This is definitely a bug - sorry if I should have labelled it as such at filing (not sure I have the rights to do so though)

Agreed that we should extend `MessageToByteEncoder`

A more "correct" fix would be to standardise the buffering encoders on a class that can handle the buffering for them.  But something like what you propose is certainly easier and less invasive.  I'll take a look at rustling up a candidate soon.

**Comment by belliottsmith** [comment]
Hmm, so taking a closer look at the other encoder implementations, this looks to only affect `Lz4FrameEncoder`, as none of the others seem to not explicitly buffer - they simply encode a block large enough to contain the input message.  So this might be simpler than we thought - I might keep it simple and just override the necessary methods in `MessageToByteEncoder` in only `Lz4FrameEncoder`

**Comment by normanmaurer** [comment]
@belliottsmith SGTM

**Comment by njhill** [comment]
@belliottsmith I think at least `Bzip2Encoder` is similarly affected? I agree that it would be nicer to abstract/consolidate equivalent buffering logic but in the bzip2 case at least, the buffering happens within a separate "block compressor". Having just an aggregate promise in the superclass which is passed to write only when `encode...` returns true seems like it would be very simple and address both cases.

**Comment by normanmaurer** [comment]
Another problem with the current implementation is that it does not correctly update the channel writability as it not uses `PendingWriteQueue` or `AbstractCoalescingBufferQueue` :/

**Comment by normanmaurer** [comment]
@belliottsmith did you have time to look into it ?

**Comment by belliottsmith** [comment]
Sorry @normanmaurer, O did not and since forgot.  It turns out it's harder to keep up with your regular responsibilities when you're distracted by a cute baby in your personal time.

I'd still be willing to take a look, but you might have to bear with me another few weeks.  Please don't feel like you're treading on my toes if you or somebody else wants to address it though.
